ruff_lsp/server.py

```diff
@@ -528,6 +810,9 @@ async def code_action(params: CodeActionParams) -> list[CodeAction] | None:
     # Add "Ruff: Autofix" for every fixable diagnostic.
     if settings.get("codeAction", {}).get("fixViolation", {}).get("enable", True):
         if not params.context.only or CodeActionKind.QuickFix in params.context.only:
+            document = Document.from_text_document(
+                LSP_SERVER.workspace.get_text_document(params.text_document.uri)
+            )
```

Comment:
Why are we creating the document again here when it's already constructed above at the start of the function?

This function is a generic implementation to create code actions at both source and line level. The document created above can only be used to create source level code actions as it represents either a Python file or an entire Notebook.

Here, we need a document which represents a Python file or an individual cell because we need to get the line level code action. Here, we already have the diagnostics information which is why this works as otherwise we would've been running Ruff on an individual cell.

Reply:
Do you mind adding something to this effect as a comment?

Reply:
I've simplified this part such that each branch uses the document through our custom abstraction or the one from `pygls`.